src/libraries/Microsoft.Extensions.Logging/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -0,0 +1,655 @@
+                            // An unescaped '{' or '}' found.
+                            break;
+                        }
+                    }
+                    else if (message[scanIndex] == brace)
+                    {
+                        if (brace == '}')
+                        {
+                            if (braceOccurrenceCount == 0)
+                            {
+                                // For '}' pick the first occurrence.
+                                braceIndex = scanIndex;
+                            }
+                        }
+                        else
+                        {
+                            // For '{' pick the last occurrence.
+                            braceIndex = scanIndex;
+                        }
+
+                        braceOccurrenceCount++;
+                    }
+
+                    scanIndex++;
+                }
+
+                return braceIndex;
+            }
+
+            private static int FindIndexOfAny(string message, char[] chars, int startIndex, int endIndex)
+            {
+                var findIndex = message.IndexOfAny(chars, startIndex, endIndex - startIndex);
+                return findIndex == -1 ? endIndex : findIndex;
+            }
+        }
+
+        /// <summary>
+        /// A logger class holding a bunch of logger methods.
+        /// </summary>
+        internal class LoggerClass
+        {
+            public readonly List<LoggerMethod> Methods = new ();
+            public string Namespace = string.Empty;
+            public string Name = string.Empty;
+            public string Constraints = string.Empty;
+        }
+
+        /// <summary>
+        /// A logger method in a logger class.
+        /// </summary>
+        internal class LoggerMethod
+        {
+            public readonly List<LoggerParameter> AllParameters = new ();
+            public readonly List<LoggerParameter> RegularParameters = new ();
+            public readonly Dictionary<string, string> TemplateMap = new (StringComparer.OrdinalIgnoreCase);
+            public readonly List<string> TemplateList = new ();
+            public string Name = string.Empty;
+            public string Message = string.Empty;
+            public int? Level;
+            public int EventId;
+            public string? EventName;
+            public bool IsExtensionMethod;
+            public string Modifiers = string.Empty;
+            public string LoggerField = "_logger";
+        }
+
+        /// <summary>
+        /// A single parameter to a logger method.
+        /// </summary>
+        internal class LoggerParameter
+        {
+            public string Name = string.Empty;
+            public string Type = string.Empty;
+            public bool IsLogger;
+            public bool IsException;
+            public bool IsLogLevel;
+            public bool IsEnumerable;
+            public bool IsRegular => !IsLogger && !IsException && !IsLogLevel;
```

Comment:
Does `IsRegular` really capture what this means? What is a `regular` parameter?

Reply:
`IsRegular` would mean the parameter is not going to be treated specially as an argument to `ILogger.Log`, but is instead supposed to be taken as a parameter for the template.

Reply:
Maybe `IsTemplateParameter`? Or at least add the above explanation to a comment on the property.